infobind/cache.go

```diff
@@ -0,0 +1,224 @@
+}
+
+// NewHandle create a Handle with a BindCache.
+func NewHandle() *Handle {
+	handle := &Handle{}
+	return handle
+}
+
+// Get get bindCache from a Handle.
+func (h *Handle) Get() *BindCache {
+	bc := h.bind.Load()
+
+	if bc != nil {
+		return bc.(*BindCache)
+	}
+
+	return nil
+}
+
+// LoadDiff use to load new bind info to bindCache bc.
+func (h *HandleUpdater) loadDiff(sql string, bc *BindCache) error {
+	tmp, err := h.Ctx.(sqlexec.SQLExecutor).Execute(context.Background(), sql)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	rs := tmp[0]
+	defer terror.Call(rs.Close)
+
+	fs := rs.Fields()
+	chk := rs.NewChunk()
+	for {
+		err = rs.Next(context.TODO(), chk)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if chk.NumRows() == 0 {
+			return nil
+		}
+		it := chunk.NewIterator4Chunk(chk)
+		for row := it.Begin(); row != it.End(); row = it.Next() {
+			record, err := decodeBindTableRow(row, fs)
+			if err != nil {
+				log.Errorf("row decode error %s", err)
+				continue
+			}
+			err = bc.appendNode(h.Ctx, record, h.Parser)
+			if err != nil {
+				continue
+			}
+
+			if record.UpdateTime.Compare(h.LastUpdateTime) == 1 {
+				h.LastUpdateTime = record.UpdateTime
+			}
+		}
+		chk = chunk.Renew(chk, h.Ctx.GetSessionVars().MaxChunkSize)
+	}
+}
+
+// Update update the HandleUpdater's bindCache if tidb first startup,the fullLoad is true,otherwise fullLoad is false.
+func (h *HandleUpdater) Update(fullLoad bool) error {
+	var (
+		err error
+		sql string
+	)
+	bc := h.Get()
+
+	length := defaultBindCacheSize
+	if bc != nil {
+		length = len(bc.Cache)
+	}
+
+	newBc := &BindCache{
+		Cache: make(map[string][]*BindData, length),
+	}
+
+	if bc != nil {
+		for hash, bindDataArr := range bc.Cache {
+			newBc.Cache[hash] = append(newBc.Cache[hash], bindDataArr...)
+		}
+	}
+
+	if fullLoad {
+		sql = fmt.Sprintf("select * from mysql.bind_info")
+	} else {
+		sql = fmt.Sprintf("select * from mysql.bind_info where update_time > \"%s\"", h.LastUpdateTime.String())
+	}
+	err = h.loadDiff(sql, newBc)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	h.bind.Store(newBc)
+	return nil
+}
+
+func parseSQL(sctx sessionctx.Context, parser *parser.Parser, sql string, charset string, collation string) ([]ast.StmtNode, []error, error) {
+	return parser.Parse(sql, charset, collation)
+}
+
+func decodeBindTableRow(row chunk.Row, fs []*ast.ResultField) (bindRecord, error) {
+	var value bindRecord
+
+	value.OriginalSQL = row.GetString(0)
+	value.BindSQL = row.GetString(1)
+	value.Db = row.GetString(2)
+	value.Status = row.GetInt64(3)
+	value.CreateTime = row.GetTime(4)
+	value.UpdateTime = row.GetTime(5)
+	value.Charset = row.GetString(6)
+	value.Collation = row.GetString(7)
+
+	return value, nil
+}
+
+func (b *BindCache) appendNode(sctx sessionctx.Context, value bindRecord, sparser *parser.Parser) error {
+	hash := parser.Digest(value.OriginalSQL)
+	if value.Status == 0 {
+		if bindArr, ok := b.Cache[hash]; ok {
+			if len(bindArr) == 1 {
```

Comment:
There is no need to specially handle this case. Just check if the arr length is 0 after line 197 is enough.